### ISSUE TITLE
[Git Bash] Impossible to run a bash script from Windows Explorer if the path/name contains non-ASCII characters

### ISSUE THREAD

**Issue by TestPolygon** [git-for-windows/git#2189]
### Description
I cannot properly run the .sh script **by clicking** on it if the path to the script contains non-ASCII character(s). The script does not work at all (I see only immediately closing of a console.).

I have Windows 10, 64-bit, Git-2.21.0 64-bit.

For example:
The content of the script:
```bash
#!/bin/bash
echo "Hello there!"
sleep 3
```
I can properly run the script **in the console** (git-bash.exe) in any way.

But, if I run the script **from the Windows' Explorer by double-clicking** on it:
- If the path contains only ASCII characters: everything is OK.
- If the path contains any **non**-ASCII character: the script (the console) closes immediately after the start, the script does not do any work.

Examples of the pathes:
Good:
- `C:\script.sh`
- `C:\#Bash\New folder\script.sh`
- `C:\#Bash\Pokmon.sh`

"Bad":
- `C:\#Bash\Pokémon.sh` Diacritic
- `C:\#Bash\Catálogo\script.sh` ES
- `C:\#Bash\Новая папка\script.sh` RU
- `C:\#Bash\New folder\脚本.sh` ZN
- `C:\#Bash\Surrogate Pair 💧\script.sh` Emoji

**I would like to be able to run scripts by double-clicking if the path contains non-ASCII characters**, at least characters from BMP (Basic Multilingual Plane).

**Comment by PhilipOakley** [comment]
Probably the place to start looking is the install scripts in the build-extra repo to see how the 'double clicking' is set up. I couldn't tell (personal ignorance ;-) if you are using the default windows mechanism for file associations etc to get the script to run, or if it was via a Git for Windows specific action mechanism (a right click option).

As a guess you are getting some character conversion issue between all the different apps/OSes/layers, such that you have, say, a windows characters being passed as if it is utf-8 or some such confusion.

**Comment by TestPolygon** [comment]
### Step to reproduce
1. You need Windows (I think any 10, 8, 7)
2. Create a bash script
3. Install git (git-bash included) the latest or not version from the site
4. After installing you can run the script from Windows' Explorer
5. **If the path to the script contains non-ASCII characters it will not work.**

**Comment by TestPolygon** [comment]
I can run the script from the console in any case:
If the script is located in `C:\#Bash\New Folder\💧\` and its name is `water 💧.sh`:
I can open bash.exe here using the context menu and run the script (even if the name of the script contains non-ASCII chars). It works.


```bash
User@User-PC MINGW64 /c/#Bash/New Folder/рџ’§                                                    
$ ./water\ ^Xрџ’§.sh
Hello there!
```
I see a little not presentable view of the output of the console, but the script works.
For name typing I use auto-complete (tab).

**Comment by TestPolygon** [comment]
The ability to run bash scripts from Windows' Explorer is very useful and convenient. In this case a bash script behave yourself like a usual program(`.exe` file). Especially it's convenient when you use drag and drop mechanism to pass arguments to the script.

**Comment by dscho** [comment]
@TestPolygon sorry for the long wait. This should be fixed in the next version of Git for Windows. To make sure, please download the 64-bit package from https://dev.azure.com/johannes-schindelin/git/_build/results?buildId=4912&view=artifacts, unpack `usr\bin\msys-2.0.dll` into your `C:\Program Files\Git` (this should overwrite an existing file, you might want to back it up in case of unforeseen complications).

**Comment by TestPolygon** [comment]
>https://dev.azure.com/johannes-schindelin/git/_build/results?buildId=4912&view=artifacts

drive.google.com, dropbox.com, mega.nz, uploadir.com or any else?

**Comment by rimrul** [comment]
@dscho that link requires a microft login. The download is probably only accessible to you.

**Comment by dscho** [comment]
>  that link requires a microsoft login

Well, boo.


> > https://dev.azure.com/johannes-schindelin/git/_build/results?buildId=4912&view=artifacts
> 
> drive.google.com, dropbox.com, mega.nz or any other file sharing service?

In the meantime, it made it into the Git for Windows SDK. Try this: https://github.com/git-for-windows/git-sdk-64/raw/master/usr/bin/msys-2.0.dll

**Comment by dscho** [comment]
There also has been a new [snapshot](https://wingit.blob.core.windows.net/files/index.html) in the meantime, if it is easier for you to run the Git for Windows installer.

**Comment by TestPolygon** [comment]
> There also has been a new [snapshot](https://wingit.blob.core.windows.net/files/index.html)

**Running** by double click works well now, but **by drag and drop does not work well.**

### How to reproduce
1. Use the bash script with the follow code:
```bash
#!/bin/bash

# Show args count
args_count=$#
echo "Args count: $args_count"

# List of sorted args
printf "%s\n" "$(printf "%s\n" "$@" | sort -n)"
sleep 2
```

2. Drap and drop any file on this `.sh` file.

If bash script has "bad" character(s) in its name or/and path, the script does not work at all when you drag and drop file(s) on it.

When only the files (that you drag and drop on the script) have "bad" character(s) in its name or/and path it _seems_ to work.

**Comment by dscho** [comment]
I tried this with https://wingit.blob.core.windows.net/files/Git-prerelease-2.24.0.windows.2.14.g8715ee2266-64-bit.exe and it worked for me with the file name `ähm.sh`... but yes, I can reproduce with the "name" you provided.

Not sure when I will get to debug this, though.

**Comment by TestPolygon** [comment]
> it worked for me with the file name `ähm.sh`

It doesn't work on my machine. 

> I can reproduce with the "name" you provided.

It's an example set of "bad" chars. The existence of any (one) chars of this set in the name or path of the script will break the work.

**Comment by rimrul** [comment]
> > it worked for me with the file name `ähm.sh`
> 
> It doesn't work on my machine.

Probably locale dependent. While umlauts aren't ASCII, they are on a buch of DOS/Windows codepages.(850,858,859,865,1252,...)

**Comment by dscho** [comment]
After being buried in my inbox for a long time, I stumbled over this again. At this stage I have to admit defeat: I won't be able to set aside enough time to investigate this issue properly.

I do not know whether this issue still reproduces with `éá脚本💧.sh`, but if it does, and if anybody here is willing to debug this (it's not really easy, but with a bit of tenacity it is totally doable), I can at least offer my assistance in [the MSYS2 runtime](https://github.com/git-for-windows/git/wiki/Building-msys2-runtime) with tons of inserted debug messages and then diagnosing the issue all the way to a fix.

**Comment by dscho** [comment]
This issue actually came back from my "Done" pile, as its bug fix is the reason that I cannot deploy the MSYS2 runtime right now.

I did verify that the bug fix that Git for Windows carried for the past 6 years is not actually necessary, double-clicking scripts with Unicode characters works just fine.

What still does not work is that drag-drop thing, and it irked me that I had no idea why, so I finally sat down and properly investigated this. Claude Opus helped make sense of 70MB worth of Process Monitor output.

tl;dr: this is not a bug in Git for Windows or the MSYS2 runtime. It is a limitation of the standard Windows Shell [DropHandler](https://learn.microsoft.com/en-us/windows/win32/shell/how-to-create-drop-handlers) (`{60254CA5-953B-11CF-8C96-00AA00B8708C}` from `shell32.dll`), which mangles non-ASCII paths through the system ANSI code page. Fixing it requires shipping a custom [COM DLL](https://github.com/msysgit/Git-Cheetah/blob/master/explorer/factory.c#L11-L19), similar to what Python did.

## **What the Procmon trace shows**

First, `git-bash.exe` itself handles Unicode paths just fine. Invoking it directly (e.g. `Start-Process git-bash.exe '--no-cd "éá脚本💧.sh"'`) works for both ASCII and non-ASCII targets: mintty appears, the script runs.

The problem is that Explorer's DropHandler never even gets that far in the failing case. I copied the `"éá脚本💧.sh"` file to `a123.sh`, started [Procmon](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) and then dragged each script onto the other:

- **Working case** (drag onto e.g. `a123.sh`): Explorer spawns `DllHost.exe` (COM Surrogate), which launches `git-bash.exe --no-cd "a123.sh" "éá脚本💧.sh"`. Full chain works.
- **Failing case** (drag onto `éá脚本💧.sh`): Explorer spawns `DllHost.exe`, and then nothing. `git-bash.exe` is never launched.

The smoking gun: in the failing case, Explorer first accesses the Unicode filename successfully (via wide-char APIs), but then ~0.5 seconds later the DropHandler code tries to access the same path and gets `NAME INVALID`. The path was garbled through ANSI conversion. Even in the working case (ASCII target), the _dragged file_ argument shows up as `éá????.sh` in process creation events: the CJK and emoji characters got replaced with `?` because they are not representable in my system's ANSI code page (e.g. CP1252 on Western Windows systems).

## **Root cause**

The registered DropHandler `{60254CA5-953B-11CF-8C96-00AA00B8708C}` is a [standard Windows Shell component](https://learn.microsoft.com/en-us/answers/questions/2664481/wsh-vbs-drophandler-doesnt-support-unicode). It converts file paths through the system ANSI code page when constructing the command line. Characters not representable in that code page (CJK, emoji, Cyrillic on a Western system, etc.) get replaced with `?`, producing an invalid path. When the _target_ `.sh` file has such characters, the path is invalid before `git-bash.exe` is ever invoked, so the launch silently fails. When the _dragged_ files have such characters, `git-bash.exe` is launched but receives garbled arguments.

## **This is the same bug Python had**

CPython solved this exact problem by shipping their own Unicode-aware DropHandler COM DLL (`pyshellext.amd64.dll`), registered under CLSID `{BEA218D2-6950-497B-9434-61683EC065FE}`. See [the original commit by Steve Dower](https://github.com/python/cpython/commit/df450d1a18ba668874a2353a3870ba99c4848a75) and [the source at `PC/pyshellext.cpp`](https://github.com/python/cpython/blob/main/PC/pyshellext.cpp) (roughly 400 lines of C++). It implements [`IPersistFile`](https://learn.microsoft.com/en-us/windows/win32/api/objidl/nn-objidl-ipersistfile) (receives the target path as Unicode `LPCOLESTR`) and [`IDropTarget`](https://learn.microsoft.com/en-us/windows/win32/api/oleidl/nn-oleidl-idroptarget) (reads dropped files via [`CF_HDROP`](https://learn.microsoft.com/en-us/windows/win32/shell/clipboard#cf_hdrop) with wide-char APIs), then calls [`ShellExecute`](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecutew) with the full Unicode paths directly, avoiding the ANSI round-trip entirely. See also [the Python bug tracker discussion](https://github.com/python/cpython/issues/71656) for more context.

## **What a fix would look like**

We would need a similar [COM DLL](https://learn.microsoft.com/en-us/windows/win32/shell/how-to-create-drop-handlers) for Git for Windows, modeled on Python's `pyshellext.cpp`, that calls `git-bash.exe --no-cd <target.sh> <dropped-files>` using [`CreateProcessW`](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw). The DLL would be registered under its own CLSID, and the `sh_auto_file\ShellEx\DropHandler` registry entry would point to it instead of the broken `{60254CA5}`. It can be [built with MinGW](https://www.transmissionzero.co.uk/computing/building-dlls-with-mingw/) rather than requiring Visual Studio.

The only alternative that avoids writing a COM DLL is to enable "Use Unicode UTF-8 for worldwide language support" in Windows Region settings (which changes the system ANSI code page to UTF-8 / CP65001). That works, but it is a system-wide change with potential side effects on other software, so it is not something we can reasonably ask users to do.

I won't try to find time to write this DLL; This investigation already cost me a lot. And contributions are always very welcome in this project, the Python implementation is a good, self-contained starting point.

**Comment by AlttiRi** [comment]
If I'm correct, `CF_HDROP` works fine with dropped files with non-ANSI characters even without enabled
`"Beta: Use Unicode UTF-8 for worldwide language support"` too.

```c++
        FORMATETC feHDrop = {   CF_HDROP,       nullptr, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
        STGMEDIUM stg;

        if (SUCCEEDED(pDataObject->GetData(&feHDrop, &stg))) {
            HDROP hDrop = (HDROP)GlobalLock(stg.hGlobal);
            if (hDrop) {
                UINT count = DragQueryFileW(hDrop, 0xFFFFFFFF, nullptr, 0);
                files.reserve(count);
                for (UINT i = 0; i < count; ++i) {
                    WCHAR szPath[MAX_PATH] = {};  // 259 (the max short path string length) + 1 (`\0` char)
                    if (DragQueryFileW(hDrop, i, szPath, MAX_PATH)) {
                        files.emplace_back(m_ResolveToLongPath(szPath));
                    }
                }
                GlobalUnlock(stg.hGlobal);
            }
            ReleaseStgMedium(&stg);
        }
```

When I tested file dropping in Windows applications, the only thing which `CF_HDROP` is not supported it's **long paths**. For this case you can use a fallback to `CFSTR_SHELLIDLIST`:
```c++
        FORMATETC feSHIDL = { m_CF_SHELLIDLIST, nullptr, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };

        // ...else
        if (SUCCEEDED(pDataObject->GetData(&feSHIDL, &stg))) {
            LPIDA pIda = (LPIDA)GlobalLock(stg.hGlobal);
            if (pIda) {
                LPCITEMIDLIST pidlFolder = (LPCITEMIDLIST)((LPBYTE)pIda + pIda->aoffset[0]);
                for (UINT i = 1; i <= pIda->cidl; ++i) {
                    LPCITEMIDLIST pidlItem = (LPCITEMIDLIST)((LPBYTE)pIda + pIda->aoffset[i]);
                    LPITEMIDLIST pidlFull = ILCombine(pidlFolder, pidlItem);
                    if (pidlFull) {
                        std::wstring path = m_GetPathFromPIDL(pidlFull);
                        if (!path.empty()) {
                            files.emplace_back(std::move(path));
                        }
                        ILFree(pidlFull);
                    }
                }
                GlobalUnlock(stg.hGlobal);
            }
            ReleaseStgMedium(&stg);
        }
```

Here is more info:
https://github.com/AlttiRi/long-path-files-drag-and-drop 
Check the readme file, it's an article about long path droppings and related things.

**Comment by dscho** [comment]
@AlttiRi feel free to give it a try. I have already spent more time on this investigation than I wanted.
